demos/http/http_demo_s3_download_async/http_demo_s3_download_async.c

```diff
@@ -0,0 +1,751 @@
+static const char * pAddress = NULL;
+
+/* The host address string extracted from S3_PRESIGNED_URL. */
+static char serverHost[ MAX_HOST_ADRESS_LENGTH ];
+
+/* The length of the host address found in string S3_PRESIGNED_URL. */
+static size_t serverHostLength = 0;
+
+/**
+ * @brief Data type for request queue.
+ *
+ * In addition to sending the headers struct, we need to send the buffer it uses.
+ * The pointer to the buffer in the headers struct will be incorrect after
+ * it is received so will need to be fixed.
+ */
+typedef struct RequestItem
+{
+    HTTPRequestHeaders_t requestHeaders;
+    uint8_t headersBuffer[ USER_BUFFER_LENGTH ];
+} RequestItem_t;
+
+/**
+ * @brief Data type for response queue.
+ *
+ * In addition to sending the response struct, we need to send the buffer it uses.
+ * The pointer to the buffer in the response struct will be incorrect after
+ * it is received so will need to be fixed.
+ */
+typedef struct ResponseItem
+{
+    HTTPResponse_t response;
+    uint8_t responseBuffer[ USER_BUFFER_LENGTH ];
+} ResponseItem_t;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Connect to HTTP server with reconnection retries.
+ *
+ * @param[out] pNetworkContext The output parameter to return the created network context.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on successful connection.
+ */
+static int32_t connectToServer( NetworkContext_t * pNetworkContext );
+
+/**
+ * @brief Enqueue HTTP requests based on a specified method and path.
+ *
+ * @param[in] pHost The host name of the server.
+ * @param[in] pMethod The HTTP request method.
+ * @param[in] pPath The Request-URI to the objects of interest.
+ * @param[in] requestQueue The queue to write requests to.
+ * @param[out] requestCount The number of HTTP requests queued.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on success.
+ */
+static int requestS3ObjectFile( const char * pHost,
+                                const char * pMethod,
+                                const char * pPath,
+                                mqd_t requestQueue,
+                                size_t * requestCount );
+
+
+/**
+ * @brief Processes HTTP responses from response queue.
+ *
+ * @param[in] responseQueue The queue to read response from.
+ * @param[in] requestCount The number of HTTP responses to read.
+ */
+static void retrieveS3ObjectFile( mqd_t responseQueue,
+                                  size_t * requestCount );
+
+/**
+ * @brief Enqueue HTTP requests based on a specified method and path.
+ *
+ * @param[in] pTransportInterface The transport interface for making network calls.
+ *
+ * @return -1 on failure; PID of HTTP thread on success.
+ */
+static pid_t startHTTPThread( const TransportInterface_t * pTransportInterface );
+
+/*-----------------------------------------------------------*/
+
+static int32_t connectToServer( NetworkContext_t * pNetworkContext )
```

Comment:
It seems like connectToServer is the same for all the S3 demos. Perhaps we want to put this in utils.